src/Compilers/Core/Portable/DiagnosticAnalyzer/IArtifactProducer.cs

```diff
@@ -0,0 +1,167 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.IO;
+using System.Text;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    /// <summary>
+    /// The base interface for artifact producers that can programmatically generate additional non-code files during a
+    /// compilation.  Artifact generators only run when a compiler is invoked with the <c>generatedartifactsout</c>
+    /// parameter.
+    /// </summary>
+    /// <remarks>
+    /// Normally, an artifact generator will not need to report diagnostics.  However, it may sometimes be necessary if
+    /// errors or other issues arise during generation.  If diagnostic reporting is needed, then the same mechanism are
+    /// available as with normal analyzers.  Specifically, <see cref="SupportedDiagnostics"/> should be overridden to
+    /// state which diagnostics may be produced, and the various context `ReportDiagnostic` calls should be used to
+    /// report them.
+    /// </remarks>
+    public interface IArtifactProducer
+    {
+        /// <summary>
+        /// Returns a set of descriptors for the diagnostics that this artifact producer is capable of producing.
+        /// </summary>
+        ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; }
```

Comment:
Is there a reason we need the SupportedDiagnostics property? My understanding is that for analyzers we require it, so that we know if all diagnostics it reports are suppressed, theres no need to run the analyzer.

It seems that a producer is more like a generator in that we'd still have to run it even with suppressions in place, so there doesn't seem to be a benefit to requiring them be declared?

Reply:
Ah, because we have the wrapper type that allows us to treat the producer as an analyzer? I think we could special case that though to ignore the `SupportedDiagnostics` check when its a producer.

Reply:
i thought about that, but i got a little worried that other parts of the system might care about this.  for example, we could potentially have an invariance check that says: you tried to report a diagnostic that you didn't say you supported.  Naughty naughty, that's illegal.  This makes the contract clear and also informs users that this is supported.